File: src/test/java/com/adven/concordion/extensions/exam/PlaceholdersResolverTest.java
package com.adven.concordion.extensions.exam;

import org.concordion.api.Evaluator;
import org.joda.time.Period;
import org.junit.Test;

import static com.adven.concordion.extensions.exam.PlaceholdersResolver.resolve;
import static org.hamcrest.Matchers.is;
import static org.joda.time.LocalDateTime.now;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class PlaceholdersResolverTest {
    Evaluator eval = mock(Evaluator.class);

    @Test
    public void canUseConcordionVars() throws Exception {
        when(eval.getVariable("#value")).thenReturn(3);

        assertThat(resolve("${var.value}", eval), is("3"));
    }

    @Test
    public void canUseJsonUnitStringAliases() throws Exception {
        String expected = "${json-unit.any-string}";
        assertThat(resolve("!{any-string}", eval), is(expected));
        assertThat(resolve("!{aNy-stRiNG}", eval), is(expected));
        assertThat(resolve("!{string}", eval), is(expected));
        assertThat(resolve("!{str}", eval), is(expected));
    }

    @Test
    public void canUseJsonUnitNumberAliases() throws Exception {
        String expected = "${json-unit.any-number}";
        assertThat(resolve("!{any-number}", eval), is(expected));
        assertThat(resolve("!{aNy-nuMBeR}", eval), is(expected));
        assertThat(resolve("!{number}", eval), is(expected));
        assertThat(resolve("!{num}", eval), is(expected));
    }

    @Test
    public void canUseJsonUnitBoolAliases() throws Exception {
        String expected = "${json-unit.any-boolean}";
        assertThat(resolve("!{any-boolean}", eval), is(expected));
        assertThat(resolve("!{aNy-bOOlean}", eval), is(expected));
        assertThat(resolve("!{boolean}", eval), is(expected));
        assertThat(resolve("!{bool}", eval), is(expected));
    }

    @Test
    public void canUseJsonUnitMatcherAliases() throws Exception {
        assertThat(resolve("!{formattedAs dd.MM.yyyy}", eval),
                is("${json-unit.matches:formattedAs}dd.MM.yyyy"));
        assertThat(resolve("!{formattedAndWithin [yyyy-MM-dd][1d][1951-05-13]}", eval),
                is("${json-unit.matches:formattedAndWithin}[yyyy-MM-dd][1d][1951-05-13]"));
    }

    @Test
    public void canAddSimplePeriodToNow() throws Exception {
        String expected = now().plusDays(1).toString("dd.MM.yyyy");

        assertThat(resolve("${exam.now+[1 d]:dd.MM.yyyy}", eval), is(expected));
        assertThat(resolve("${exam.now+[1 day]:dd.MM.yyyy}", eval), is(expected));
        assertThat(resolve("${exam.now+[day 1]:dd.MM.yyyy}", eval), is(expected));
        assertThat(resolve("${exam.now+[1 days]:dd.MM.yyyy}", eval), is(expected));
    }

    @Test
    public void canAddCompositePeriodsToNow() throws Exception {
        assertThat(
                resolve(
                        "${exam.now+[1 day, 1 month]:dd.MM.yyyy}", eval
                ), is(
                        now().plus(new Period().
                                plusDays(1).plusMonths(1)).toString("dd.MM.yyyy")
                )
        );
        assertThat(
                resolve(
                        "${exam.now+[days 3, months 3, 3 years]:dd.MM.yyyy}", eval
                ), is(
                        now().plus(new Period().
                                plusDays(3).plusMonths(3).plusYears(3)).toString("dd.MM.yyyy")
                )
        );
        assertThat(
                resolve(
                        "${exam.now+[4 d, 4 M, y 4, 4 hours]:dd.MM.yyyy'T'hh}", eval
                ), is(
                        now().plus(new Period().
                                plusDays(4).plusMonths(4).plusYears(4).plusHours(4)).toString("dd.MM.yyyy'T'hh")
                )
        );
    }

    @Test
    public void canSubtractCompositePeriodsFromNow() throws Exception {
        assertThat(
                resolve(
                        "${exam.now-[1 day, 1 month]:dd.MM.yyyy}", eval
                ), is(
                        now().minusDays(1).minusMonths(1).toString("dd.MM.yyyy")
                )
        );
        assertThat(
                resolve(
                        "${exam.now-[days 3, months 3, 3 years]:dd.MM.yyyy}", eval
                ), is(
                        now().minusDays(3).minusMonths(3).minusYears(3).toString("dd.MM.yyyy")
                )
        );
        assertThat(
                resolve(
                        "${exam.now-[4 d, 4 M, y 4, 4 hours]:dd.MM.yyyy'T'hh}", eval
                ), is(
                        now().minusDays(4).minusMonths(4).minusYears(4).minusHours(4).toString("dd.MM.yyyy'T'hh")
                )
        );
    }
}

File: src/test/java/com/adven/concordion/extensions/exam/PlaceholdersResolverTest.kt
package com.adven.concordion.extensions.exam

import com.adven.concordion.extensions.exam.core.resolveJson
import com.adven.concordion.extensions.exam.core.resolveToObj
import com.adven.concordion.extensions.exam.core.toStringOr
import org.assertj.core.api.Assertions
import org.concordion.internal.OgnlEvaluator
import org.hamcrest.Matchers.`is`
import org.joda.time.LocalDateTime.fromDateFields
import org.joda.time.LocalDateTime.now
import org.joda.time.format.DateTimeFormat.forPattern
import org.junit.Assert.assertThat
import org.junit.Test
import java.util.Date
import kotlin.test.assertEquals


class PlaceholdersResolverTest {
    private val eval = OgnlEvaluator(Object())

    @Test
    fun canUseConcordionVars() {
        eval.setVariable("#value", 3)

        assertThat(eval.resolveJson("{{value}}"), `is`("3"))
        assertEquals(eval.resolveToObj("{{value}}"), 3)
    }

    @Test
    fun resolveToObj_shouldResolveFormattedConcordionVarToString() {
        val date = Date()
        eval.setVariable("#value", date)

        val expected = forPattern("dd.MM.yyyy HH:mm").print(fromDateFields(date))
        assertThat(eval.resolveToObj("{{dateFormat value \"dd.MM.yyyy HH:mm\"}}").toString(), `is`(expected))
        assertThat(eval.resolveToObj("{{dateFormat (now plus='1 d') 'dd.MM.yyyy HH:mm' minus='1 d'}}").toString(), `is`(expected))
    }

    @Test
    fun `resolveToObj should skip resolving if starts and ends with apostrophe`() {
        assertThat(eval.resolveToObj(" '{{resolve \"some\"}}'").toStringOr(""), `is`("some"))
        assertThat(eval.resolveToObj("'{{resolve \"some\"}}' ").toStringOr(""), `is`("some"))
        assertThat(eval.resolveToObj("'{{resolve \"some\"}}'").toStringOr(""), `is`("{{resolve \"some\"}}"))
    }

    @Test
    fun canUseJsonUnitMatcherAliases() {
        assertThat(
            eval.resolveJson("{{formattedAs 'dd.MM.yyyy'}}"),
            `is`("\${json-unit.matches:formattedAs}dd.MM.yyyy")
        )
        assertThat(
            eval.resolveJson("{{formattedAs 'dd.MM.yyyy HH:mm'}}"),
            `is`("\${json-unit.matches:formattedAs}dd.MM.yyyy HH:mm")
        )
        assertThat(
            eval.resolveJson("{{formattedAndWithin 'yyyy-MM-dd' '1d' '1951-05-13'}}"),
            `is`("\${json-unit.matches:formattedAndWithin}[yyyy-MM-dd][1d][1951-05-13]")
        )
    }

    @Test
    fun examDateVariables() {
        val p = "dd.MM.yyyy'T'hh:mm:ss"
        assertThat(eval.resolveJson("{{now \"$p\"}}"), `is`(now().toString(p)))
        Assertions.assertThat(eval.resolveToObj("{{now}}") as Date)
            .isCloseTo(now().toDate(), 1000)
        Assertions.assertThat(eval.resolveToObj("{{now minus='1 d'}}") as Date)
            .isCloseTo(now().minusDays(1).toDate(), 1000)
    }
}